### DISCUSSION TITLE
mini-css-extract-plugin question

### DISCUSSION THREAD

**Discussion by damiankomonski** [webpack/webpack#17359] [Q&A]
Hello,

This is my piece of my webpack configuration:
```
            {
                test: /\.js$/,
                exclude: /node_modules/,
                use: {
                    loader: "babel-loader",
                    options: {
                        presets: ["@babel/preset-env"]
                    }
                }
            },
            {
                test: /\.(png|svg|jpg|jpeg|gif)/i,
                type: "asset/resource",
                generator: {
                    outputPath: "../images/",
                    publicPath: "images/"
                }
            },
```

When I have such line in mini css extract plugin
`filename: "../css/[name].css",`

I get bad links in CSS output file.  For example
`{background-image:url(../js//images/vertical_bg.svg)`

All is good. I get normal CSS paths:
`{background-image:url(/images/vertical_bg.svg)`

How can I fix it?

**Comment by alexander-akait** [comment]
Why you use `filename: "../css/[name].css",`? Why you use `outputPath: "../fonts/",`? And etc, if you need to have relative URLs, just set `publicPath: "auto"`

**Reply by alexander-akait** [reply]
If you want to have CSS in another place, just set `filename` and `chunkFilname` for`mini-css-extract-loader`

**Reply by damiankomonski** [reply]
I used other filename because I want to have CSS in other localization. If I change filename webpack generate bad paths.... I wrote about it.

**Reply by damiankomonski** [reply]
Also if I have such configuration:
```
            {
                test: /\.(png|svg|jpg|jpeg|gif)/i,
                type: "asset/resource",
                generator: {
                    outputPath: "../images/",
                    publicPath: ""
                }
            },
```

Webpack generate me such path:
`{background-image:url(../js//horizontal_bg.svg)`

It adds unnecessary backslash "/". Of course path is still bad.

**Reply by alexander-akait** [reply]
It means you have a mistake somewhere, I strongly recommended only change `filename`, `assetFilename`, `chunkFilename` and etc, and don't touch `publicPath`s

**Comment by damiankomonski** [comment]
@alexander-akait 
I remove publicPath and set fileName and change filename, chunkFilename in mini-css-extract-plugin. Links are still bad.

My conf:
```
module.exports = {
    entry: { 
        main: "./js/index.js" 
    },
    output: {
        path: path.resolve(__dirname, "js"),
        filename: "scripts.js",
        clean: false,
        assetModuleFilename: "[name][ext]"
    },
```


Links generated by webpack:
`{background-image:url(../js/horizontal_bg.svg);`

They should be 
`{background-image:url(../images/horizontal_bg.svg);`

**Reply by alexander-akait** [reply]
You don't need `outputPath: "../fonts/"`, it was not design for such usage, set `filename: 'fonts/[hash][ext]'"` or `filename: 'images/[hash][ext]',` instead `outputPath`, avoid using `..` in filename at all, for JS you can use `filename: "javascript/scripts.js",`


For css:
```
new MiniCssExtractPlugin({
            filename: "css/[name].css",
            chunkFilename: "css/[id].css",
            ignoreOrder: false,
}),
```

**Reply by damiankomonski** [reply]
Thank you for response. So I rather should create dist/src folder as I good understand?

Could you tell me how should I deal with html files?

**Reply by damiankomonski** [reply]
Could explain also is it possible to export images which I have in HTML files?

**Reply by alexander-akait** [reply]
> Thank you for response. So I rather should create dist/src folder as I good understand?

Yes, one directory for source files, another directory for bundled files

> Could you tell me how should I deal with html files?

There is `html-loader` if you want to use HTML inside JavaScript and there is https://github.com/jantimon/html-webpack-plugin to generate HTML with scripts and links tags

**Comment by damiankomonski** [comment]
@alexander-akait 
Ok I did it. It seems to work. But I have one question yet. I have such folder structure in images:

```
images/portfolio/1/screen.png
images/portfolio/2/screen.png
images/image1.png
images/image2.png
```

I am using such code to load images in HTML files
`<%=require('./images/services_mobile.png')%>`
`<%=require('./images/portfolio/1/screen.png')%>`

Is it possible to have the same structure in dist folder, in files which webpack export? I want to webpack create the same folders when export images. 

Now I get also such error
`ERROR in   Error: Conflict: Multiple chunks emit assets to the same filename images/screen.png (chunks HtmlWebpackPlugin_0-0 and HtmlWebpackPlugin_0-0)`

Please help.

**Reply by alexander-akait** [reply]
It means you try to save in the `image` directory two different values, that is why we recommend to use `[hash]` in filenames, it prevents it.

I.e. you have `images/portfolio/1/screen.png` and `images/portfolio/2/screen.png` and then you try to save it to the `dist` directory you have the same filename for both files - `images/screen.png`

**Reply by damiankomonski** [reply]
Ok, I understand, but can webpack create the same folder structure?

**Reply by damiankomonski** [reply]
It solves my problem, but could you tell me if there is some way to create the same folder structure as I have in "src" folder? I am speaking about images.

**Reply by damiankomonski** [reply]
@alexander-akait 
I have one question yet. I insert images in html files in such way
`<%=require('./images/services_mobile.png')%>`

What if I have other file and I want to insert the same image? How should I do this?

**Comment by damiankomonski** [comment]
I have problem when I am using HTMLWebpackPlugin. Problem is that when I change something in index.js file webpack doesnt export images. When I run webpack first time, it exports all assets. When I change something in js file, webpack remove images. I have watch turn on.

My configuration:
```
...
```

**Reply by damiankomonski** [reply]
@alexander-akait please help

**Reply by alexander-akait** [reply]
It is an old bug https://github.com/jantimon/html-webpack-plugin/issues/1639, I want to fix it soon

**Reply by damiankomonski** [reply]
:/ so how can I work? What should I do?

**Reply by damiankomonski** [reply]
I want load images and work with HTML files, export them.

**Reply by alexander-akait** [reply]
Set `cache: false` for `HtmlWebpackPlugin`

**Reply by damiankomonski** [reply]
It doesnt work. After some change in javascript file, images arent export.

```
        new HtmlWebpackPlugin({
            filename: "index.html",
            template: "src/index.html",
            inject: "body",
            scriptLoading: "defer",
            cache: false
        }),
```

**Reply by alexander-akait** [reply]
What do you mean `exported`?

**Reply by damiankomonski** [reply]
Images which I include in html files

`<img alt="" data-src="<%=require('./images/services_mobile.png')%>" data-retina="<%=require('./images/services_mobile@2x.png')%>" class="lazy">`

**Comment by damiankomonski** [comment]
@alexander-akait 
I added this
```
            {
                test: /\.html$/,
                type: "asset/resource",
                generator: {
                    filename: "[name][ext]",
                },
            },
            {
                test: /\.html$/i,
                loader: "html-loader",
                options: {
                    sources: {
                      list: [
                        "...",
                        {
                          tag: "div",
                          attribute: "data-src",
                          type: "src"
                        },
                        {
                            tag: "img",
                            attribute: "data-src",
                            type: "src"
                          },
                      ],
                    },
                  },
            },
```

and remove all HtmlWebpackPlugin. I get error:
```
ERROR in ./src/index.html
Module build failed (from ./node_modules/html-loader/dist/cjs.js):
URIError: URI malformed
```

https://webpack.js.org/loaders/html-loader/#export-into-html-files

**Reply by damiankomonski** [reply]
Ok I solved this error. Bad link in index.html . With this configuration, webpack doesnt export html files. Why?

**Reply by alexander-akait** [reply]
What do you mean export HTML files?

**Reply by damiankomonski** [reply]
Generate html files. Create them.

**Reply by damiankomonski** [reply]
I get nervous with it :/ I asked about simple things, why cant you tell me how can I make it? ... It seems to be faster using Gulp.js than set it....

**Reply by alexander-akait** [reply]
I'm not your personal assistant, sorry. You can use `HtmlWebpackPlugin` for HTML files generation, I don't undestand your problem and what do you want, so I can't answer.

You use `<%=require('./images/services_mobile.png')%>` and want to parse this files as HTML? How webpack can do it? How webpack can know what is `<%= %>`? I can't undestand what you try to setup, because no one does it - you can use or template syntax and use `require` or use pure HTML and `html-loader` to handle `src` and etc attributes

**Reply by alexander-akait** [reply]
In this case don't use:
```html
<img alt="" data-src="<%=require('./images/services_mobile.png')%>" data-retina="<%=require('./images/services_mobile@2x.png')%>" class="lazy">
```

just write:
```html
<img alt="" data-src="./images/services_mobile.png" data-retina="./images/services_mobile@2x.png" class="lazy">
```

You don't need to `require(...)`, html-loader handle all attributes

**Reply by damiankomonski** [reply]
Thanks for your help.

Could you write something more how does it work? html-loader and HTMLWebpackPlugin? How does HTMLWebpackPlugin knows that I am using html-loader? I have to import html file in entry point, yes?

**Reply by damiankomonski** [reply]
I only write that I tried use extract-loader as written in the documentation but I got errors. It seems to doesnt work html-loader with extract-loader.

I also tried use such configuration:
```
            {
                test: /\.html$/,
                type: "asset/resource",
                generator: {
                  filename: "[name][ext]",
                },
            },
            {
                test: /\.html$/i,
                loader: "html-loader",
                options: {
                    sources: {
                      list: [
                        "...",
                        {
                            tag: "div",
                            attribute: "data-src",
                            type: "src"
                        },
                        {
                            tag: "img",
                            attribute: "data-src",
                            type: "src"
                        },
                      ],
                    },
                  },
            }
```

But it generate weird html with such code:
```
// Imports
import ___HTML_LOADER_GET_SOURCE_FROM_IMPORT___ from "../node_modules/html-loader/dist/runtime/getUrl.js";
var ___HTML_LOADER_IMPORT_0___ = new URL("./images/favicon.ico", import.meta.url);
var ___HTML_LOADER_IMPORT_1___ = new URL("./images/slider_bg.jpg", import.meta.url);
var ___HTML_LOADER_IMPORT_2___ = new URL("./images/services_mobile.png", import.meta.url);
var ___HTML_LOADER_IMPORT_3___ = new URL("./images/services_web.png", import.meta.url);
var ___HTML_LOADER_IMPORT_4___ = new URL("./images/services_branding.png", import.meta.url);
var ___HTML_LOADER_IMPORT_5___ = new URL("./images/services_marketing.png", import.meta.url);
```

So the only option seems to be to use html-loader and HTMLWebpackPlugin. But how does it work exactly, I don't know right now.

**Reply by alexander-akait** [reply]
But what is the problem use `html-loader` and `HTMLWebpackPlugin`?

> Could you write something more how does it work? html-loader and HTMLWebpackPlugin? How does HTMLWebpackPlugin knows that I am using html-loader? I have to import html file in entry point, yes?

No, `HTMLWebpackPlugin` try to compile file to HTML using any loader which registered for `html`, so webpack apply `html-loader` and then `HTMLWebpackPlugin` run it on Node.js side and extract HTML

**Reply by damiankomonski** [reply]
There is no problem to use it. It is ok.

I have to tell HTMLWebpackPlugin which html should export:
```
new HtmlWebpackPlugin({
            filename: "services.html",
            template: "src/services.html",
            inject: "body",
            scriptLoading: "defer"
        }),
```
and pass template. I wonder how it works that webpack exports html files and change paths in appropriate.

I have one question yet, if I can. How can I work with pdf files? I have such code in html file
`<a href="./../images/portfolio/26_abc/1.pdf" target="_blank" style="color: inherit">`

html-loader doesnt load this file. How can I do this? Of course I have other <a> tags with normal links. Could you give me a hint?

**Reply by alexander-akait** [reply]
Webpack (i.e. html-loader) doesn't handle `href`, there is a lint of all supported things https://github.com/webpack-contrib/html-loader#sources and there is how you can improve it https://github.com/webpack-contrib/html-loader#object

**Reply by damiankomonski** [reply]
Hello, I write such code to my configuration:

```
            {
                test: /\.(pdf)/i,
                type: "asset/resource",
                generator: {
                    filename: "pdf/[name][hash][ext]"
                }
            },
            {
                test: /\.html$/i,
                loader: "html-loader",
                options: {
                    sources: {
                      list: [
                        "...",
                        {
                            tag: "div",
                            attribute: "data-src",
                            type: "src"
                        },
                        {
                            tag: "div",
                            attribute: "data-retina",
                            type: "src"
                        },
                        {
                            tag: "img",
                            attribute: "data-src",
                            type: "src"
                        },
                        {
                            tag: "img",
                            attribute: "data-retina",
                            type: "src"
                        },
                        {
                            tag: "a",
                            attribute: "href",
                            type: "src",
                            filter: (tag, attribute, attributes, resourcePath) => {
                                attributes.forEach((element) => {
                                    console.log(element);
                                    if(element.name === "data-hrefpdf" && element.value.length > 0){
                                        return true;
                                    } else {
                                        return false;
                                    }
                                })
                            }
                        }
                      ],
                    },
                  },
            }
```

Unfortunately webpack doesnt export pdf file and change path in html file.

My html code
`<a href="./../images/portfolio/26/1.pdf" target="_blank" style="color: inherit" data-hrefpdf="./../images/portfolio/26/1.pdf">`

Could you tell me why? As I good understand if filter function return true path is being processed.

**Reply by damiankomonski** [reply]
Sorry my mistake. It works. I return true/false from foreach callback....

**Reply by damiankomonski** [reply]
Hmm there is a lot of <a> elements in my files, maybe it slow downs webpack. But now it works :)

**Reply by damiankomonski** [reply]
I also exported image files in javascript file and path is correct locally and on the server. When I use it in some javascript plugin. I thought that it will be problem with path. How it works, I dont know ;p Maybe it takes some hostname?

`import image1 from "./images/1.png"`

**Reply by alexander-akait** [reply]
We have `publicPath: 'auto'`, i.e. automatically found your script URL and recreate a valid URL to your asset

**Reply by damiankomonski** [reply]
Could you tell me if it is possible to have the directory structure in dist folder for images? I tried with [path] but it doesnt work :/ What should I change? For example I dont want use [hash], I want to have the same directory structure.

I have one question yet:
`<meta property="og:image" content="./images/avatar.png" />`

I have such meta tag, I want to webpack generate full path for this image, www.link.com/images/avatar.png . How should I do this?

If you can, please help me.

**Reply by alexander-akait** [reply]
> Could you tell me if it is possible to have the directory structure in dist folder for images? I tried with [path] but it doesnt work :/ What should I change? For example I dont want use [hash], I want to have the same directory structure.

Yes, but I don't reccommend, because you can have file overlapping and it will create long URLs in files so you will have bigger files.

> I have such meta tag, I want to webpack generate full path for this image, [www.link.com/images/avatar.png](http://www.link.com/images/avatar.png) . How should I do this?

Webpack generates relative URLs by default, if you need absolute URLs, you need to use full URL to your website in the `publicPath` option

**Comment by webdiscus** [comment]
@damiankomonski 

to [keep the same folder structure](https://github.com/webdiscus/html-bundler-webpack-plugin#recipe-keep-folder-structure-assets) use the `filename` as the function, e.g.:

```js
{
  test: /\.(png|svg|jpg|jpeg|gif)/i,
  type: "asset/resource",
  generator: {
    filename: ({ filename }) => {
      const srcPath = 'src/assets/images'; // path to source images
      const regExp = new RegExp(`[\\\\/]?(?:${path.normalize(srcPath)}|node_modules)[\\\\/](.+?)$`);
      const assetPath = path.dirname(regExp.exec(filename)[1].replace('@', '').replace(/\\/g, '/'));

      return `images/${assetPath}/[name][ext]`;
    },
  }
},
```

The destructed `filename` argument is the source file. It can be an absolute or a relative.
